url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/DNSClient.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2015-2016 the original author or authors
+ *
+ * This software is licensed under the Apache License, Version 2.0,
+ * the GNU Lesser General Public License version 2 or later ("LGPL")
+ * and the WTFPL.
+ * You may choose either license to govern your use of this software only
+ * upon the condition that you accept all of the terms of either
+ * the Apache License 2.0, the LGPL 2.1+ or the WTFPL.
+ */
+package de.measite.minidns;
+
+import de.measite.minidns.dnsserverlookup.AndroidUsingExec;
+import de.measite.minidns.dnsserverlookup.AndroidUsingReflection;
+import de.measite.minidns.dnsserverlookup.DNSServerLookupMechanism;
+import de.measite.minidns.dnsserverlookup.HardcodedDNSServerAddresses;
+import de.measite.minidns.dnsserverlookup.UnixUsingEtcResolvConf;
+import de.measite.minidns.util.MultipleIoException;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.logging.Level;
+
+/**
+ * A minimal DNS client for SRV/A/AAAA/NS and CNAME lookups, with IDN support.
+ * This circumvents the missing javax.naming package on android.
+ */
+public class DNSClient extends AbstractDNSClient {
+
+    static final List<DNSServerLookupMechanism> LOOKUP_MECHANISMS = new ArrayList<>();
+
+    static {
+        addDnsServerLookupMechanism(AndroidUsingExec.INSTANCE);
+        addDnsServerLookupMechanism(AndroidUsingReflection.INSTANCE);
+        addDnsServerLookupMechanism(HardcodedDNSServerAddresses.INSTANCE);
+        addDnsServerLookupMechanism(UnixUsingEtcResolvConf.INSTANCE);
+    }
+
+    private final Set<String> nonRaServers = Collections.newSetFromMap(new ConcurrentHashMap<String, Boolean>(4));
+
+    private boolean askForDnssec = false;
+    private boolean disableResultFilter = false;
+
+    /**
+     * Create a new DNS client using the global default cache.
+     */
+    public DNSClient() {
+        super();
+    }
+
+    public DNSClient(DNSCache dnsCache) {
+        super(dnsCache);
+    }
+
+    @Override
+    protected DNSMessage.Builder newQuestion(DNSMessage.Builder message) {
+        message.setRecursionDesired(true);
+        message.getEdnsBuilder().setUdpPayloadSize(dataSource.getUdpPayloadSize()).setDnssecOk(askForDnssec);
+        return message;
+    }
+
+    @Override
+    public DNSMessage query(DNSMessage.Builder queryBuilder) throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.